Duplicate union across namespaces creates problems, but only in Gradle task

**Describe the bug**
When two schemas have a union on a field with the same name, an error occurs: "Duplicate generated class names found with different schema files. This usually indicates that the same full-name is used across multiple avro schema files but with different content."

This _only_ occurrs when running the Gradle task, not in `KotlinGenerator`. KotlinGenerator just chooses to generate the _last_ occurrence of the Union class `FieldArrayUnion` pointing only to `ns2`, so no valid classes were generated for `ns1`.

This happens because the classes are generated in folders named after their namespaces, but the _union_ sealed interface are generated in the _root_. So the two unions with the same name either clash or one overwrites the other.

I found this when using the task on our project's schema files which have a similar setup.

**Stack trace**
Gradle output:
```
Duplicate generated class names found with different schema files. This usually indicates that the same full-name is used across multiple avro schema files but with different content. Please check your input schema files:
FieldArrayUnion has been generated differently from the following schemas: /home/sander/work/prive/avro4k/example/kotlin-plugin/src/main/avro/main-ns1.avsc, /home/sander/work/prive/avro4k/example/kotlin-plugin/src/main/avro/main-ns2.avsc.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateAvroKotlinSources'.
> Duplicate generated class names found with different schema files. This usually indicates that the same full-name is used across multiple avro schema files but with different content. Please check your input schema files:
  FieldArrayUnion has been generated differently from the following schemas: /home/sander/work/prive/avro4k/example/kotlin-plugin/src/main/avro/main-ns1.avsc, /home/sander/work/prive/avro4k/example/kotlin-plugin/src/main/avro/main-ns2.avsc.
```

**To Reproduce**
Code to reproduce the behavior:
Two similar schemas with a different namespace:
```json
{
    "type": "record",
    "name": "Main",
    "namespace": "ns1",
    "fields": [
        {
            "name": "field",
            "type": {
                "type": "array",
                "items": [
                    {
                        "type": "record",
                        "name": "A",
                        "fields": [
                            {
                                "name": "field",
                                "type": "int"
                            }
                        ]
                    },
                    {
                        "type": "record",
                        "name": "B",
                        "fields": [
                            {
                                "name": "field",
                                "type": "int"
                            }
                        ]
                    }
                ]
            }
        }
    ]
}
```
---
```json
{
    "type": "record",
    "name": "Main",
    "namespace": "ns2",
    "fields": [
        {
            "name": "field",
            "type": {
                "type": "array",
                "items": [
                    {
                        "type": "record",
                        "name": "A",
                        "fields": [
                            {
                                "name": "field",
                                "type": "int"
                            }
                        ]
                    },
                    {
                        "type": "record",
                        "name": "B",
                        "fields": [
                            {
                                "name": "field",
                                "type": "int"
                            }
                        ]
                    }
                ]
            }
        }
    ]
}
```
The generated source files are structured like this:
```
.
├── FieldArrayUnion.kt
├── ns1
│   ├── A.kt
│   ├── B.kt
│   └── Main.kt
└── ns2
    ├── A.kt
    ├── B.kt
    └── Main.kt
```

See my branch for the full example: https://github.com/sanderv/avro4k/tree/bugfix/duplicate-union

**Expected behavior**
I would expect the union to be created in the samen namespace (folder) as their surrounding classes.

Then a question:  Why isn't the `SourcesGeneration` function `verifyNoNameClash` part of KotlinGenerator? That's where most of your test are.


**Desktop (please complete the following information):**
Using the project sources and versions

**Additional context**
I'm happy to help you fix this :)

--- a/kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt
+++ b/kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt
@@ -129,11 +129,30 @@ public class KotlinGenerator(
                 listOf(recordType.toFileSpec(schema.name.space)) +
                     // generate nested types
                     schema.fields.flatMap { field ->
-                        // the union schema is already generated as a subtype of the record, no need to generate it again. However, we need to generate the types used in the union
-                        if (field.schema is AvroSchema.UnionSchema) {
-                            field.schema.nonNullTypes.flatMap { generateNestedKotlinClasses(it, potentialAnonymousClassName, emptyMap()) }
-                        } else {
-                            generateNestedKotlinClasses(field.schema, field.name.toPascalCase(), mapOf(schema.asClassName() to recordType))
+                        // the union schema, and union in a map/array schema, is already generated as a subtype of the record, no need to generate it again. However, we need to generate the types used in the union
+                        val currentRecord = schema.asClassName() to recordType
+                        when (field.schema) {
+                            is AvroSchema.UnionSchema -> {
+                                field.schema.types.flatMap {
+                                    generateNestedKotlinClasses(it, potentialAnonymousClassName, mapOf(currentRecord))
+                                }
+                            }
+
+                            is AvroSchema.ArraySchema if field.schema.elementSchema is AvroSchema.UnionSchema -> {
+                                field.schema.elementSchema.types.flatMap {
+                                    generateNestedKotlinClasses(it, potentialAnonymousClassName, mapOf(currentRecord))
+                                }
+                            }
+
+                            is AvroSchema.MapSchema if field.schema.valueSchema is AvroSchema.UnionSchema -> {
+                                field.schema.valueSchema.types.flatMap {
+                                    generateNestedKotlinClasses(it, potentialAnonymousClassName, mapOf(currentRecord))
+                                }
+                            }
+
+                            else -> {
+                                generateNestedKotlinClasses(field.schema, field.name.toPascalCase(), mapOf(currentRecord))
+                            }
                         }
                     }
             }
@@ -226,11 +245,30 @@ public class KotlinGenerator(
                     // generate nested types
                     listOf(recordType.toFileSpec(schema.name.space)) +
                         schema.fields.flatMap { field ->
-                            // the union schema is already generated as a subtype of the record, no need to generate it again. However, we need to generate the types used in the union
-                            if (field.schema is AvroSchema.UnionSchema) {
-                                field.schema.types.flatMap { generateNestedKotlinClasses(it, potentialAnonymousBaseName, generatedRecords + (schema.asClassName() to recordType)) }
-                            } else {
-                                generateNestedKotlinClasses(field.schema, field.name.toPascalCase(), generatedRecords + (schema.asClassName() to recordType))
+                            // the union schema, and union in a map/array schema, is already generated as a subtype of the record, no need to generate it again. However, we need to generate the types used in the union
+                            val currentRecord = schema.asClassName() to recordType
+                            when (field.schema) {
+                                is AvroSchema.UnionSchema -> {
+                                    field.schema.types.flatMap {
+                                        generateNestedKotlinClasses(it, potentialAnonymousBaseName, generatedRecords + currentRecord)
+                                    }
+                                }
+
+                                is AvroSchema.ArraySchema if field.schema.elementSchema is AvroSchema.UnionSchema -> {
+                                    field.schema.elementSchema.types.flatMap {
+                                        generateNestedKotlinClasses(it, potentialAnonymousBaseName, generatedRecords + currentRecord)
+                                    }
+                                }
+
+                                is AvroSchema.MapSchema if field.schema.valueSchema is AvroSchema.UnionSchema -> {
+                                    field.schema.valueSchema.types.flatMap {
+                                        generateNestedKotlinClasses(it, potentialAnonymousBaseName, generatedRecords + currentRecord)
+                                    }
+                                }
+
+                                else -> {
+                                    generateNestedKotlinClasses(field.schema, field.name.toPascalCase(), generatedRecords + currentRecord)
+                                }
                             }
                         }
                 } else {
@@ -489,11 +527,23 @@ public class KotlinGenerator(
             }
             .addTypes(
                 schema.fields.mapNotNull { field ->
-                    if (field.schema is AvroSchema.UnionSchema && !field.schema.isSimpleNullableType) {
-                        val unionBaseName = field.name.toPascalCase()
-                        generateSealedInterface(field.schema, unionNameFormatter(unionBaseName), unionBaseName)
-                    } else {
-                        null
+                    val unionBaseName = field.name.toPascalCase()
+                    when (field.schema) {
+                        is AvroSchema.ArraySchema if field.schema.elementSchema is AvroSchema.UnionSchema && !field.schema.elementSchema.isSimpleNullableType -> {
+                            generateSealedInterface(field.schema.elementSchema, unionNameFormatter(arrayNameFormatter(unionBaseName)), unionBaseName)
+                        }
+
+                        is AvroSchema.MapSchema if field.schema.valueSchema is AvroSchema.UnionSchema && !field.schema.valueSchema.isSimpleNullableType -> {
+                            generateSealedInterface(field.schema.valueSchema, unionNameFormatter(mapNameFormatter(unionBaseName)), unionBaseName)
+                        }
+
+                        is AvroSchema.UnionSchema if !field.schema.isSimpleNullableType -> {
+                            generateSealedInterface(field.schema, unionNameFormatter(unionBaseName), unionBaseName)
+                        }
+
+                        else -> {
+                            null
+                        }
                     }
                 }
             )

--- a/kotlin-generator/src/test/expected-sources/duplicates/ns1/A.kt
+++ b/kotlin-generator/src/test/expected-sources/duplicates/ns1/A.kt
@@ -1,0 +1,17 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.internal.AvroGenerated
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"A","namespace":"ns1","fields":[{"name":"field","type":"int"}]}""")
+public data class A(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/duplicates/ns1/B.kt
+++ b/kotlin-generator/src/test/expected-sources/duplicates/ns1/B.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"B","namespace":"ns1","fields":[{"name":"field","type":"int"}]}""")
+public data class B(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/duplicates/ns1/Main.kt
+++ b/kotlin-generator/src/test/expected-sources/duplicates/ns1/Main.kt
@@ -1,0 +1,37 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.OptIn
+import kotlin.collections.List
+import kotlin.collections.emptyList
+import kotlin.jvm.JvmInline
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"Main","namespace":"ns1","fields":[{"name":"field","type":{"type":"array","items":[{"type":"record","name":"A","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","fields":[{"name":"field","type":"int"}]}]}}]}""")
+public data class Main(
+    public val `field`: List<FieldArrayUnion> = emptyList(),
+) {
+    @Serializable
+    @AvroGenerated("""[{"type":"record","name":"A","namespace":"ns1","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","namespace":"ns1","fields":[{"name":"field","type":"int"}]}]""")
+    public sealed interface FieldArrayUnion {
+        @JvmInline
+        @Serializable
+        public value class ForA(
+            public val `value`: A,
+        ) : FieldArrayUnion
+
+        @JvmInline
+        @Serializable
+        public value class ForB(
+            public val `value`: B,
+        ) : FieldArrayUnion
+    }
+}

--- a/kotlin-generator/src/test/expected-sources/duplicates/ns2/A.kt
+++ b/kotlin-generator/src/test/expected-sources/duplicates/ns2/A.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"A","namespace":"ns2","fields":[{"name":"field","type":"int"}]}""")
+public data class A(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/duplicates/ns2/B.kt
+++ b/kotlin-generator/src/test/expected-sources/duplicates/ns2/B.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"B","namespace":"ns2","fields":[{"name":"field","type":"int"}]}""")
+public data class B(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/duplicates/ns2/Main.kt
+++ b/kotlin-generator/src/test/expected-sources/duplicates/ns2/Main.kt
@@ -1,0 +1,37 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.OptIn
+import kotlin.collections.List
+import kotlin.collections.emptyList
+import kotlin.jvm.JvmInline
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"Main","namespace":"ns2","fields":[{"name":"field","type":{"type":"array","items":[{"type":"record","name":"A","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","fields":[{"name":"field","type":"int"}]}]}}]}""")
+public data class Main(
+    public val `field`: List<FieldArrayUnion> = emptyList(),
+) {
+    @Serializable
+    @AvroGenerated("""[{"type":"record","name":"A","namespace":"ns2","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","namespace":"ns2","fields":[{"name":"field","type":"int"}]}]""")
+    public sealed interface FieldArrayUnion {
+        @JvmInline
+        @Serializable
+        public value class ForA(
+            public val `value`: A,
+        ) : FieldArrayUnion
+
+        @JvmInline
+        @Serializable
+        public value class ForB(
+            public val `value`: B,
+        ) : FieldArrayUnion
+    }
+}

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns1/A.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns1/A.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"A","namespace":"ns1","fields":[{"name":"field","type":"int"}]}""")
+public data class A(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns1/B.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns1/B.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"B","namespace":"ns1","fields":[{"name":"field","type":"int"}]}""")
+public data class B(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns1/Main.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns1/Main.kt
@@ -1,0 +1,37 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.OptIn
+import kotlin.collections.List
+import kotlin.collections.emptyList
+import kotlin.jvm.JvmInline
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"Main","namespace":"ns1","fields":[{"name":"field","type":{"type":"array","items":[{"type":"record","name":"A","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","fields":[{"name":"field","type":"int"}]}]}}]}""")
+public data class Main(
+    public val `field`: List<FieldArrayUnion> = emptyList(),
+) {
+    @Serializable
+    @AvroGenerated("""[{"type":"record","name":"A","namespace":"ns1","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","namespace":"ns1","fields":[{"name":"field","type":"int"}]}]""")
+    public sealed interface FieldArrayUnion {
+        @JvmInline
+        @Serializable
+        public value class ForA(
+            public val `value`: A,
+        ) : FieldArrayUnion
+
+        @JvmInline
+        @Serializable
+        public value class ForB(
+            public val `value`: B,
+        ) : FieldArrayUnion
+    }
+}

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns2/A.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns2/A.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"A","namespace":"ns2","fields":[{"name":"field","type":"int"}]}""")
+public data class A(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns2/B.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns2/B.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"B","namespace":"ns2","fields":[{"name":"field","type":"int"}]}""")
+public data class B(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns2/Main.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-array/ns2/Main.kt
@@ -1,0 +1,37 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.OptIn
+import kotlin.collections.List
+import kotlin.collections.emptyList
+import kotlin.jvm.JvmInline
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"Main","namespace":"ns2","fields":[{"name":"field","type":{"type":"array","items":[{"type":"record","name":"A","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","fields":[{"name":"field","type":"int"}]}]}}]}""")
+public data class Main(
+    public val `field`: List<FieldArrayUnion> = emptyList(),
+) {
+    @Serializable
+    @AvroGenerated("""[{"type":"record","name":"A","namespace":"ns2","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","namespace":"ns2","fields":[{"name":"field","type":"int"}]}]""")
+    public sealed interface FieldArrayUnion {
+        @JvmInline
+        @Serializable
+        public value class ForA(
+            public val `value`: A,
+        ) : FieldArrayUnion
+
+        @JvmInline
+        @Serializable
+        public value class ForB(
+            public val `value`: B,
+        ) : FieldArrayUnion
+    }
+}

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns1/A.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns1/A.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"A","namespace":"ns1","fields":[{"name":"field","type":"int"}]}""")
+public data class A(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns1/B.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns1/B.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"B","namespace":"ns1","fields":[{"name":"field","type":"int"}]}""")
+public data class B(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns1/Main.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns1/Main.kt
@@ -1,0 +1,38 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns1
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.OptIn
+import kotlin.String
+import kotlin.collections.Map
+import kotlin.collections.emptyMap
+import kotlin.jvm.JvmInline
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"Main","namespace":"ns1","fields":[{"name":"field","type":{"type":"map","values":[{"type":"record","name":"A","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","fields":[{"name":"field","type":"int"}]}]}}]}""")
+public data class Main(
+    public val `field`: Map<String, FieldMapUnion> = emptyMap(),
+) {
+    @Serializable
+    @AvroGenerated("""[{"type":"record","name":"A","namespace":"ns1","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","namespace":"ns1","fields":[{"name":"field","type":"int"}]}]""")
+    public sealed interface FieldMapUnion {
+        @JvmInline
+        @Serializable
+        public value class ForA(
+            public val `value`: A,
+        ) : FieldMapUnion
+
+        @JvmInline
+        @Serializable
+        public value class ForB(
+            public val `value`: B,
+        ) : FieldMapUnion
+    }
+}

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns2/A.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns2/A.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"A","namespace":"ns2","fields":[{"name":"field","type":"int"}]}""")
+public data class A(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns2/B.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns2/B.kt
@@ -1,0 +1,19 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.Int
+import kotlin.OptIn
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"B","namespace":"ns2","fields":[{"name":"field","type":"int"}]}""")
+public data class B(
+    public val `field`: Int,
+)

--- a/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns2/Main.kt
+++ b/kotlin-generator/src/test/expected-sources/nested-union-in-map/ns2/Main.kt
@@ -1,0 +1,38 @@
+@file:OptIn(
+    InternalAvro4kApi::class,
+    ExperimentalAvro4kApi::class,
+)
+
+package ns2
+
+import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
+import com.github.avrokotlin.avro4k.`internal`.AvroGenerated
+import kotlin.OptIn
+import kotlin.String
+import kotlin.collections.Map
+import kotlin.collections.emptyMap
+import kotlin.jvm.JvmInline
+import kotlinx.serialization.Serializable
+
+@Serializable
+@AvroGenerated("""{"type":"record","name":"Main","namespace":"ns2","fields":[{"name":"field","type":{"type":"map","values":[{"type":"record","name":"A","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","fields":[{"name":"field","type":"int"}]}]}}]}""")
+public data class Main(
+    public val `field`: Map<String, FieldMapUnion> = emptyMap(),
+) {
+    @Serializable
+    @AvroGenerated("""[{"type":"record","name":"A","namespace":"ns2","fields":[{"name":"field","type":"int"}]},{"type":"record","name":"B","namespace":"ns2","fields":[{"name":"field","type":"int"}]}]""")
+    public sealed interface FieldMapUnion {
+        @JvmInline
+        @Serializable
+        public value class ForA(
+            public val `value`: A,
+        ) : FieldMapUnion
+
+        @JvmInline
+        @Serializable
+        public value class ForB(
+            public val `value`: B,
+        ) : FieldMapUnion
+    }
+}

--- a/kotlin-generator/src/test/resources/nested-union-in-array/main-ns1.avsc
+++ b/kotlin-generator/src/test/resources/nested-union-in-array/main-ns1.avsc
@@ -1,0 +1,35 @@
+{
+    "type": "record",
+    "name": "Main",
+    "namespace": "ns1",
+    "fields": [
+        {
+            "name": "field",
+            "type": {
+                "type": "array",
+                "items": [
+                    {
+                        "type": "record",
+                        "name": "A",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    },
+                    {
+                        "type": "record",
+                        "name": "B",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    }
+                ]
+            }
+        }
+    ]
+}

--- a/kotlin-generator/src/test/resources/nested-union-in-array/main-ns2.avsc
+++ b/kotlin-generator/src/test/resources/nested-union-in-array/main-ns2.avsc
@@ -1,0 +1,35 @@
+{
+    "type": "record",
+    "name": "Main",
+    "namespace": "ns2",
+    "fields": [
+        {
+            "name": "field",
+            "type": {
+                "type": "array",
+                "items": [
+                    {
+                        "type": "record",
+                        "name": "A",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    },
+                    {
+                        "type": "record",
+                        "name": "B",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    }
+                ]
+            }
+        }
+    ]
+}

--- a/kotlin-generator/src/test/resources/nested-union-in-map/main-ns1.avsc
+++ b/kotlin-generator/src/test/resources/nested-union-in-map/main-ns1.avsc
@@ -1,0 +1,35 @@
+{
+    "type": "record",
+    "name": "Main",
+    "namespace": "ns1",
+    "fields": [
+        {
+            "name": "field",
+            "type": {
+                "type": "map",
+                "values": [
+                    {
+                        "type": "record",
+                        "name": "A",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    },
+                    {
+                        "type": "record",
+                        "name": "B",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    }
+                ]
+            }
+        }
+    ]
+}

--- a/kotlin-generator/src/test/resources/nested-union-in-map/main-ns2.avsc
+++ b/kotlin-generator/src/test/resources/nested-union-in-map/main-ns2.avsc
@@ -1,0 +1,35 @@
+{
+    "type": "record",
+    "name": "Main",
+    "namespace": "ns2",
+    "fields": [
+        {
+            "name": "field",
+            "type": {
+                "type": "map",
+                "values": [
+                    {
+                        "type": "record",
+                        "name": "A",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    },
+                    {
+                        "type": "record",
+                        "name": "B",
+                        "fields": [
+                            {
+                                "name": "field",
+                                "type": "int"
+                            }
+                        ]
+                    }
+                ]
+            }
+        }
+    ]
+}

